timeboost/s3_storage.go

```diff
@@ -0,0 +1,196 @@
+	f.String(prefix+".object-prefix", DefaultS3StorageServiceConfig.ObjectPrefix, "prefix to add to S3 objects")
+	f.String(prefix+".region", DefaultS3StorageServiceConfig.Region, "S3 region")
+	f.String(prefix+".secret-key", DefaultS3StorageServiceConfig.SecretKey, "S3 secret key")
+	f.Duration(prefix+".upload-interval", DefaultS3StorageServiceConfig.UploadInterval, "frequency at which batches are uploaded to S3")
+	f.Int(prefix+".max-batch-size", DefaultS3StorageServiceConfig.MaxBatchSize, "max size of uncompressed batch in bytes to be uploaded to S3")
+	f.Int(prefix+".max-db-rows", DefaultS3StorageServiceConfig.MaxDbRows, "when the sql db is very large, this enables reading of db in chunks instead of all at once which might cause OOM")
+}
+
+type S3StorageService struct {
+	stopwaiter.StopWaiter
+	config       *S3StorageServiceConfig
+	client       s3client.FullClient
+	sqlDB        *SqliteDatabase
+	bucket       string
+	objectPrefix string
+}
+
+func NewS3StorageService(config *S3StorageServiceConfig, sqlDB *SqliteDatabase) (*S3StorageService, error) {
+	client, err := s3client.NewS3FullClient(config.AccessKey, config.SecretKey, config.Region)
+	if err != nil {
+		return nil, err
+	}
+	return &S3StorageService{
+		config:       config,
+		client:       client,
+		sqlDB:        sqlDB,
+		bucket:       config.Bucket,
+		objectPrefix: config.ObjectPrefix,
+	}, nil
+}
+
+func (s *S3StorageService) Start(ctx context.Context) {
+	s.StopWaiter.Start(ctx, s)
+	s.CallIteratively(s.uploadBatches)
+}
+
+func (s *S3StorageService) uploadBatch(ctx context.Context, batch []byte, fistRound uint64) error {
+	compressedData, err := util.CompressGzip(batch)
+	if err != nil {
+		return err
+	}
+	now := time.Now()
+	key := fmt.Sprintf("%svalidated-timeboost-bids/%d/%02d/%02d/%d.csv.gzip", s.objectPrefix, now.Year(), now.Month(), now.Day(), fistRound)
+	putObjectInput := s3.PutObjectInput{
+		Bucket: aws.String(s.bucket),
+		Key:    aws.String(key),
+		Body:   bytes.NewReader(compressedData),
+	}
+	if _, err = s.client.Upload(ctx, &putObjectInput); err != nil {
+		return err
+	}
+	return nil
+}
+
+// downloadBatch is only used for testing purposes
+func (s *S3StorageService) downloadBatch(ctx context.Context, key string) ([]byte, error) {
+	buf := manager.NewWriteAtBuffer([]byte{})
+	if _, err := s.client.Download(ctx, buf, &s3.GetObjectInput{
+		Bucket: aws.String(s.bucket),
+		Key:    aws.String(key),
+	}); err != nil {
+		return nil, err
+	}
+	return util.DecompressGzip(buf.Bytes())
+}
+
+func csvRecordSize(record []string) int {
+	size := len(record) // comma between fields + newline
+	for _, entry := range record {
+		size += len(entry)
+	}
+	return size
+}
+
+func (s *S3StorageService) uploadBatches(ctx context.Context) time.Duration {
+	bids, round, err := s.sqlDB.GetBids(s.config.MaxDbRows)
+	if err != nil {
+		log.Error("Error fetching validated bids from sql DB", "round", round, "err", err)
+		return 0
```

Comment:
Don't return zero seconds on the error cases, give it 5 or 10 seconds so that `CallIteratively` waits and we don't fill up the logs with error messages if there is some unrecoverable issue.

Reply:
I agree and fixed this. 

Up on further thinking I realized there was an issue on how I was handling errors regarding deleting of bids from SqlDB which couldve led to duplicate upload of batches! 
To fix this I've implemented tracking of last failed delete round and then prioritizing deleting of these uploaded bids upon retry.